[CELEBORN-1056] Introduce Rest API of listing dynamic configuration

SteNicholas opened the pull request:
### What changes were proposed in this pull request?

Introduce Rest API of listing dynamic configuration `/listDynamicConfigs` to list the dynamic configs. The result of `/listDynamicConfigs` is as follows:

```
=========================== Dynamic Configuration ============================
celeborn.master.ha.ratis.raft.server.snapshot.auto.trigger.threshold          100000
celeborn.worker.flusher.buffer.size                                           64k
=========================== SYSTEM ============================
celeborn.master.ha.ratis.raft.server.snapshot.auto.trigger.threshold          200000
celeborn.worker.flusher.buffer.size                                           128k
=========================== TENANT ============================
=========================== Tenant: tenantId1 ============================
celeborn.master.ha.ratis.raft.server.snapshot.auto.trigger.threshold          300000
celeborn.worker.flusher.buffer.size                                           256k
=========================== TENANT_USER ============================
=========================== Tenant: tenantId1, Name: user1 ============================
celeborn.master.ha.ratis.raft.server.snapshot.auto.trigger.threshold          400000
celeborn.worker.flusher.buffer.size                                           512k
```

### Why are the changes needed?

Celeborn supports dynamic configuration with `ConfigService` at present. It's recommend to introduce Rest API of dynamic configuration management.

### Does this PR introduce _any_ user-facing change?

- Introduce Rest API of listing dynamic configuration: `/listDynamicConfigs?level=[system|tenant|tenant_user]&tenant=tenantId1&name=user1`.

### How was this patch tested?

- `HttpUtilsSuite#CELEBORN-1056: Introduce Rest API of listing dynamic configuration`

AngersZhuuuu commented:
`Introduce Rest API of listing dynamic configuration management `? Since we may support setting restful API later

AngersZhuuuu commented on service/src/main/scala/org/apache/celeborn/server/common/http/HttpEndpoint.scala:
```diff
@@ -39,6 +42,22 @@ case object Conf extends HttpEndpoint {
     service.getConf
 }
 
+case object DynamicConfig extends HttpEndpoint {
```
ListDynamicConfig?

AngersZhuuuu commented on service/src/main/scala/org/apache/celeborn/server/common/http/HttpEndpoint.scala:
```diff
@@ -39,6 +42,22 @@ case object Conf extends HttpEndpoint {
 }
 
+case object DynamicConfig extends HttpEndpoint {
+  override def path: String = "/dynamicConfig"
```
ditto

AngersZhuuuu commented on service/src/main/scala/org/apache/celeborn/server/common/HttpService.scala:
```diff
@@ -27,15 +33,77 @@ abstract class HttpService extends Service with Logging {
+      val sb = new StringBuilder
+      sb.append("=========================== Dynamic Configuration ============================\n")
+      level.split(",").filter(_.nonEmpty).foreach {
+        configLevel =>
```
```
level.split(",").filter(_.nonEmpty).foreach { configLevel =>
   ....
}
```

AngersZhuuuu commented on service/src/main/scala/org/apache/celeborn/server/common/HttpService.scala:
```diff
@@ -27,15 +33,77 @@ abstract class HttpService extends Service with Logging {
+      sb.append("=========================== Dynamic Configuration ============================\n")
+      level.split(",").filter(_.nonEmpty).foreach {
+        configLevel =>
+          {
```
Can remove this `{}` pair

AngersZhuuuu commented on service/src/main/scala/org/apache/celeborn/server/common/HttpService.scala:
```diff
@@ -27,15 +33,77 @@ abstract class HttpService extends Service with Logging {
+                if (tenant.isEmpty) configService.listRawTenantConfigsFromCache().asScala
+                else tenant.split(",").map { tenantId =>
+                  configService.getRawTenantConfigFromCache(tenantId)
+                }.toList
```
```
if (tenant.isEmpty) {
    configService.listRawTenantConfigsFromCache().asScala
} else {
    tenant.split(",").map { tenantId =>  configService.getRawTenantConfigFromCache(tenantId) }.toList
}
```

AngersZhuuuu commented on service/src/main/scala/org/apache/celeborn/server/common/http/HttpEndpoint.scala:
```diff
@@ -39,6 +42,22 @@ case object Conf extends HttpEndpoint {
+      parameters.getOrElse("LEVEL", ConfigLevel.values().mkString(","))
+        .toUpperCase(Locale.ROOT).trim,
+      parameters.getOrElse("TENANT", "").trim,
+      parameters.getOrElse("NAME", "").trim)
```
I think
1. If set level with system level, we only show system level config
2. If set level with tenant level, without tenant id, should list all tenant level config
3. If set level with tenant level with tenant id, should only show this tenant id 's config
4. If set level with tenant user level, without setting id and name, should show all user level
5. If set level with tenant user level, with setting tenant id, should this tenant id's all user level config
6. If set level with tenant user level, with setting id and name, should show only this user's user level config
7. If not set level, but set tenant id or name, should follow above each level's behavior.

AngersZhuuuu reviewed (commented):
Also better provide a real restful result's string in desc?

SteNicholas commented:
@AngersZhuuuu, thanks for detailed review. I have addressed above comments and updated the title and description of this pull request. PTAL.

SteNicholas commented:
@AngersZhuuuu, thanks for detailed review. I have addressed above comments and updated the title and description of this pull request. PTAL. cc @RexXiong.

RexXiong commented on service/src/main/scala/org/apache/celeborn/server/common/HttpService.scala:
```diff
@@ -27,15 +33,107 @@ abstract class HttpService extends Service with Logging {
+    }.mkString("\n")
+  }
+
+  private def tenantUserConfigs(tenant: String, name: String): String = {
```
IMO and from the parameter of tenantUserConfigs, we should only support single tenant/single tenant with username to simplify and avoid the combinations of tenantId and usernames, if tenant/name are both empty we list all configs by config_level

RexXiong reviewed (approved):
LGTM, thanks。

SteNicholas commented:
Merge to main(v0.5.0).
